### ISSUE TITLE
undefined method `fields' for nil:NilClass

### ISSUE THREAD

**Issue by u007** [rails/rails#18027]
im encountering a very odd behavior that does not happen often.
it happen after running several days.
on line /home/mypathn/rails/mypath/app/models/user.rb:105:
def role? (role)
    !!self.roles.where(:name => role.to_s ).first #line 105 here
end

i do not see any issues with this as user class is not nil , but the underlying query causes nil exception with fields variable...
does anyone know where could be the issue?

undefined method `fields' for nil:NilClass
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/mysql2_adapter.rb:233:in `exec_query'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/mysql2_adapter.rb:240:in `select'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/database_statements.rb:24:in `select_all'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `block in select_all'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/query_cache.rb:83:in `cache_sql'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `select_all'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/querying.rb:39:in `find_by_sql'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/relation.rb:611:in `exec_queries'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/association_relation.rb:19:in `exec_queries'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/relation.rb:493:in `load'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/relation.rb:238:in `to_a'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/relation/finder_methods.rb:474:in `find_nth_with_limit'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/relation/finder_methods.rb:468:in `find_nth'
/home/mypathn/rails/mypath/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/relation/finder_methods.rb:132:in `first'
/home/mypathn/rails/mypath/app/models/user.rb:105:in `role?'

**Comment by u007** [comment]
hi,

sorry , ive done the a mistake in my code that causes the error,
thanks for your test script that help me pointed out my error
thanks again :)

**Comment by pbrisbin** [comment]
@u007 I'm getting the same error from a relatively straight forward looking `Model.where(...).first` call. It is intermittent and has only been seen in production. Could you tell me what mistake you found in your code that caused this, in case it helps me track down my issue? Thanks.

Feel free to email if you'd rather not spam this closed Issue: pbrisbin at gmail

**Comment by u007** [comment]
hi pbrishbin
i cant really recall,
but it has something to do with this function
user = User.create!
assert !user.role?("admin"), "should not have admin role"

in model
def role? (role)
 -    !!self.roles.where(:name => role.to_s ).first #this line caused the error
 +    !self.roles.where(:name => role.to_s ).first.nil?

end

**Comment by u007** [comment]
im getting the same error and it crashed my rails instance:
and i do not have the lead on where is the error:

Processing by ChatController#live_feed as JSON
  Parameters: {"_lang"=>"en", "_uuid"=>"4af9f465-19b6-34eb-9f80-bb2d2e99574a"}
chat-livefeed: undefined method `fields' for nil:NilClass/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/mysql2_adapter.rb:233:in `exec_query'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/mysql2_adapter.rb:240:in `select'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/database_statements.rb:24:in `select_all'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `block in select_all'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/query_cache.rb:83:in `cache_sql'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `select_all'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/database_statements.rb:30:in `select_one'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/database_statements.rb:35:in `select_value'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/relation/finder_methods.rb:298:in `exists?'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.8/lib/active_record/validations/uniqueness.rb:25:in `validate_each'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activemodel-4.1.8/lib/active_model/validator.rb:167:in `block in validate'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activemodel-4.1.8/lib/active_model/validator.rb:164:in `each'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activemodel-4.1.8/lib/active_model/validator.rb:164:in `validate'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activesupport-4.1.8/lib/active_support/callbacks.rb:447:in `public_send'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activesupport-4.1.8/lib/active_support/callbacks.rb:447:in `block in make_lambda'
/home/xxxn/rails/xxx/vendor/bundle/ruby/2.1.0/gems/activesupport-4.1.8/lib/active_support/callbacks.rb:184:in `call'

**Comment by rafaelfranca** [comment]
The only way of getting this error is if the connection adapter `execute` method returns `nil`. We need a way to reproduce this issue or we'll consider it invalid.

**Comment by u007** [comment]
ive added a inspect on it, will  update you when it crash again, 
now seems good..
it seems that its difficult for us to pin point the source of the problem...
is it due to a custom lambda function been used to call the statement?
is there anyway to find which piece of code calling the lambda function?

thanks

**Comment by rafaelfranca** [comment]
If you can't always reproduce it, I believe it is a database problem, or a problem in your setup. Closing since we can't reproduce.

**Comment by u007** [comment]
im getting the error,
but when i queried in mysql, it return no record,
but why is it causing the rails instance to crash?

console output:
exec: "SELECT  `user_attrs`.* FROM `user_attrs`  WHERE (mobile='+2189-218999999' OR clean_mobile='+218999999')  ORDER BY `user_attrs`.`id` ASC LIMIT 1", name: UserAttr Load
undefined method `fields' for nil:NilClass

 def exec_query(sql, name = 'SQL', binds = [])
        Rails.logger.info "exec: #{sql.inspect}, name: #{name}"
        result = execute(sql, name) # this part return nil...
        ActiveRecord::Result.new(result.fields, result.to_a)
      end

**Comment by u007** [comment]
Hi

I think it has todo with mysql connection pool.
I was using faye on rails.. N there is some weird behaviours that caused to have many cinnections.
After separated the faye onto another rails instance with no mysql .. The problem seems to be no longer happening. Does this relate to the ongoing bugs that caused mysql connection issues with rails livestream?

Is there a way for me to do a unittest to set pool to one and simulate simultaneously connection requests to trigger this error?

**Comment by jodosha** [comment]
@rafaelfranca Hello, I'm experiencing this problem in testing code. After the first run it works fine, then it raises this error randomly.
It looks like a shared state, but I'm not using spring, as the spec is run via `bundler exec rspec path/to/spec.rb`.

I'm on Rails 4.1.10, pg 0.18.2 (PostgreSQL 9.4.1) and marginalia 1.3.0.

It happens even for trivial lines like:

```ruby
class User < ActiveRecord::Base
  def self.by_email!(email)
    where(email: email).first!
  end
end
```

This isn't a new application, until yesterday everything worked fine, but that finder is new.

**Comment by jodosha** [comment]
@rafaelfranca I have an update. On subsequent runs of the same spec, I get the following error:

```
ActiveRecord::StatementInvalid:
       PG::UnableToSend: another command is already in progress
[REDACTED QUERY]
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/postgresql_adapter.rb:832:in `send_query_prepared'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/postgresql_adapter.rb:832:in `block in exec_cache'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract_adapter.rb:378:in `block in log'
     # /Users/luca/.gem/ruby/2.2.2/gems/activesupport-4.1.10/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract_adapter.rb:372:in `log'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/postgresql_adapter.rb:831:in `exec_cache'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/postgresql/database_statements.rb:138:in `exec_query'
     # /Users/luca/.gem/ruby/2.2.2/gems/marginalia-1.3.0/lib/marginalia.rb:56:in `exec_query_with_marginalia'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/postgresql_adapter.rb:954:in `select'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract/database_statements.rb:24:in `select_all'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `block in select_all'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract/query_cache.rb:83:in `cache_sql'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `select_all'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract/database_statements.rb:30:in `select_one'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/connection_adapters/abstract/database_statements.rb:35:in `select_value'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/relation/finder_methods.rb:298:in `exists?'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/associations/collection_association.rb:387:in `include?'
     # /Users/luca/.gem/ruby/2.2.2/gems/activerecord-4.1.10/lib/active_record/associations/collection_proxy.rb:864:in `include?'

# ...
```

The offending line is:

```ruby
@team.users.include?(user)
```

It looks like the shared state (mentioned in the previous comment) is the database. Probably some process remains stuck after the first run and causes this problem.

**Comment by u007** [comment]
jodosha - that dont look like the same issue

**Comment by u007** [comment]
btw, ive updates,
it turned out that after ive moved faye to another process,
and upgraded my eventmachine to 1.0.7,
the error no longer happen,
but instead, im getting some other eventmachine related error

im not sure how that have caused issue on this package, is it due to concurrency issue with mysql pool?

thanks

**Comment by jodosha** [comment]
@u007 The initial error for me is the same and it happens in `exec_query` but for PG adapter.

**Comment by stereoscott** [comment]
Sorry to stir up an old issue, but I wanted to share our experience with this error. We use PG Bouncer for connection pooling to postgres. When we add a new server to our formation, that new server will usually run into this issue once and then it goes away. It always occurs with the same query, which executes within

**Comment by Sinjo** [comment]
@stereoscott Do you have a unique constraint on that table by any chance? We've run into many instances of this bug now (having recently switched on transaction pooling in PgBouncer) and every instance we've seen coincided with the violation of a unique constraint.

**Comment by muffy** [comment]
I've noticed it happens almost any time we run rails migrations on our db.

**Comment by stereoscott** [comment]
@Sinjo Yes we do we do have a unique constraint on the table. Attempting to `rescue ActiveRecord::RecordNotUnique => e` around the query that triggers the exception doesn't make a difference. 

Looking at the method that raises the exception, I see this:

```ruby
result = without_prepared_statement?(binds) ? exec_no_cache(sql, name, binds) :
                                                        exec_cache(sql, name, binds)
types = {}
fields = result.fields
```

I know that PGBouncer can only run without prepared statements, so it may be the case that pg_bouncer has nothing to do with it and it's just the exec_no_cache method that exhibits the error (or a combination of the two).

I found more discussion here:

https://groups.google.com/forum/#!topic/ruby-pg/QaN4UaSvkyc

**Comment by Sinjo** [comment]
@stereoscott We (I was working with a couple of colleagues) got to the bottom this in the end, and posted [this thread](https://groups.google.com/forum/#!topic/rubyonrails-core/vwCQidpdD58) to the Rails mailing list.

#22101 made it into 4.2.5, which gets rid of the worst of this problem. I'd love some input on that mailing list thread. This is a bigger problem than #22101.

**Comment by kingdonb** [comment]
For whatever it's worth, I was in the process of upgrading a Rails 4.2.x app to Rails 5.0, diligently paying attention and fixing each deprecation warning that popped up in my tests and ran into this again.  @Sinjo says it was fixed in 4.2.5, well maybe it came back in 5.x?  Anyway struggled with this for a while, tried different versions of pg (as I read that pg 1.0.0 wasn't going to work with less than some 5.1.x...) tried switching back and forth between chromedriver-helper and webdriver, as it seemed to be some nondeterministic issue, I am not using a PGbouncer, tried all this all to no avail, then finally conceded that 5.0 is a dead branch and proceeded with the rest of the upgrade.

Upgraded to 5.1, upgraded pg to `'~> 1.0'`, everything seems fine and I haven't seen this come back again.

> The only way of getting this error is if the connection adapter execute method returns nil

This is exactly what was happening, I inserted pry statements into my gemset to find out, although I could not provide any steps to reproduce it, happens perhaps when a web service request is answered at the same time as a database connection in another thread is firing?

Seems non-deterministic, like I said the issue seems to have passed after finishing the upgrades to at least the most current 5.1 patch release and latest gemset for my app.

Thanks reporters for making notes here, I really thought I was going crazy until I found this issue.
